client/src/extension.ts

```diff
@@ -5,46 +5,145 @@
 
 import * as fs from "fs";
 import * as path from "path";
-import { promisify } from "util";
-import { workspace, ExtensionContext, commands } from "vscode";
-
-let exists = promisify(fs.exists);
+import { window, ProgressLocation, workspace, ExtensionContext, commands } from "vscode";
+import * as semver from 'semver';
 
 import {
   LanguageClient,
   LanguageClientOptions,
   ServerOptions,
-  TransportKind
 } from "vscode-languageclient";
 
+import { log, run, options } from './utils';
+import { getEsyStatus } from './esy-utils';
+import { ToolNotFound, OpamNotFound, EsyNotFound, OpamOnWindowsError } from './errors';
+import { promisify } from "util";
+
 let client: LanguageClient;
 
-async function isEsyProject() {
+async function getCommandForWorkspace(context): Promise<{ command: string, args: string[] }> {
+
   let root = workspace.rootPath;
-  if (await exists(path.join(root, "package.json"))) {
-    return true;
-  } else if (await exists(path.join(root, "esy.json"))) {
-    return true;
-  } else {
-    return false;
-  }
-}
+  let esyStatus = await getEsyStatus(root);
+  if (esyStatus.isProject) {
+
+    // All npm and opam projects are valid esy projects too! Picking
+    // the right package manager is important - we don't want to run
+    // `esy` for a user who never intended to. Example: bsb/npm
+    // users. Similarly, opam users wouldn't want prompts to run
+    // `esy`. Why is prompting `esy i` even necessary in the first
+    // place? `esy ocamlmerlin-lsp` needs projects to install/solve deps
+
+    let manifestFile = esyStatus.rootPackageConfigPath;
+
+    // Looking for manifest file. If it is not a json file, it's an opam project
+    if (/\.json$/.test(manifestFile)) {
+
+      let manifest;
+      try {
+        manifest = fs.readFileSync(manifestFile).toString();
+      } catch (e) {
+        return Promise.reject(e);

```

Comment:
This is the same as `throw e`, no? We could remove `try { ... } catch (e) { .. }` then.